### PR TITLE
[hackathon] ci: Only generate cuda tests for cuda configurations

### PR DESCRIPTION
**Pull request by seemethere**
Signed-off-by: Eli Uriegas <eliuriegas@fb.com>

### CODE REVIEW

**Comment by seemethere** [comment]
Looks like this change reduces the run time of cuda configurations by at least half:

* With change: 1h6m: https://app.circleci.com/pipelines/github/pytorch/pytorch/297552/workflows/804c30d7-731b-4eec-af1c-38d046c57069/jobs/12198915
* Without change: 2h24m: https://app.circleci.com/pipelines/github/pytorch/pytorch/297339/workflows/2f52ece2-0c94-4928-9138-8120eb4e7270/jobs/12185405

File: torch/testing/_internal/common_device_type.py [hunk 1]
```diff
@@ -476,15 +476,24 @@ def instantiate_device_type_tests(generic_test_class, scope, except_for=None, on
     generic_members = set(generic_test_class.__dict__.keys()) - set(empty_class.__dict__.keys())
     generic_tests = [x for x in generic_members if x.startswith('test')]
 
+    # Derive defaults from environment variables if available, default is still none
+    # Usage:
+    # export PYTORCH_TESTING_DEVICE_ONLY_FOR=cuda,cpu
+    # export PYTORCH_TESTING_DEVICE_EXCEPT_FOR=xla
```
Comment (walterddr) [inline torch/testing/_internal/common_device_type.py:482]: good to explain all the current supported device_type, or how to find them (i myself find it hard to expect what is going to be run using EXCEPT_FOR=xla)

Comment (seemethere) [reply torch/testing/_internal/common_device_type.py:482]: For PyTorch proper there's only actually two device types that are available, `cuda,cpu`

I think however there's documentation on the top of this file that outlines all of the possible device types

**Comment by seemethere** [comment]
Looks like this also reduces windows test time execution, although by a less dramatic amount
* With change 1h3m: https://app.circleci.com/pipelines/github/pytorch/pytorch/297801/workflows/8cbd22ea-2a92-4d3e-9d85-00645238eab4/jobs/12208441
* Without change 1h46m: https://app.circleci.com/pipelines/github/pytorch/pytorch/297808/workflows/6f3963b6-d091-4425-a045-3dc7f3d21cd2/jobs/12209325

**Review by walterddr — APPROVED** [review]
looks good to me

File: .jenkins/pytorch/test.sh [hunk 2]
```diff
@@ -23,6 +23,13 @@ if [[ "$BUILD_ENVIRONMENT" == *coverage* ]]; then
   export PYTORCH_COLLECT_COVERAGE=1
 fi
 
+if [[ "$BUILD_ENVIRONMENT" == *cuda* ]]; then
+  # Used so that only cuda specific versions of tests are generated
+  # mainly used so that we're not spending extra cycles testing cpu
+  # devices on expensive gpu machines
+  export PYTORCH_TESTING_ONLY_FOR="cuda"
```
Comment (ailzhang) [inline .jenkins/pytorch/test.sh:30]: This env doesn't match the one we get in python file :P

**Comment by kshitij12345** [comment]
Unable to test locally without ENV variables. See #55664

**Pull request closed without merging**